### PR TITLE
[gym, deliver, actions] allow gym to create .pkg for macOS apps and Catalyst apps for App Store upload and new `build_mac_app` action

### LINKED ISSUE #15719: Add support for Catalyst
#### Motivation Behind Feature

Apple advertises Catalyst as the easiest way to bring iPad apps to macOS. https://developer.apple.com/mac-catalyst/

#### Feature Description

If the `SUPPORTS_MACCATALYST` build setting is set to `YES`, fastlane could allow a user to choose whether they want to build for iOS or macOS, resulting in either an ipa file or an app bundle.
Ideally, snapshot would also allow snapshotting on macOS, adhering to https://help.apple.com/app-store-connect/#/devd274dd925, but that could be a follow-up request (see #11092).

#### Alternatives or Workarounds

Currently, it's possible to perform some actions (like `run_tests`) on Catalyst-enabled projects by passing `platform=macOS,arch=x86_64,variant=Mac Catalyst` as `destination`.
However, this fails for other actions like `build_ios_app` which assume that an iOS project always produces an ipa or app bundle. For example, https://github.com/fastlane/fastlane/blob/a934ee6dc08c62098f0f28f9a91cefae8d568bd6/gym/lib/gym/runner.rb#L23 fails with `The generated archive is invalid`.

### PR DESCRIPTION
**Pull request by joshdholtz**
Fixes #15719

### Gym
- Able to "build" Mac apps but was not able to package them for App Store to pass off to `deliver`
  - This caused a little confusion since both `gym` and `deliver` docs said that Mac was supported
  - `deliver` actually supports uploading of a `.pkg` file but `gym` was not able to make a `.pkg` file
- No intuitive way of supporting Catalyst apps due to `gym` not knowing which platform to create
- dsym lane context was build specifically for finding iOS dsym but not macOS dsym

### Deliver
- Can upload both `.pkg` and `.ipa` but would favor `.ipa` no matter which platform was selected (which is a problem for Catalyst because `.ipa` and `.pkg` artifacts will be in the project directory)

### Actions
- `build_app` is now the main driver
- `gym` aliases `build_app`
- `build_ios_app` is subclass of `build_app`
  - Filters out macOS specific options
  - Defaults `catalyst_platform` to `ios`
- `build_mac_app` is subclass of `build_app`
  - Filters out iOS specific options
  - Defaults `catalyst_platform` to `macos`

### macOS
- `gym` has two new options for Mac App Store signing
  - `mac_app_installer_cert_name`
    - Value can auto detected by `Gym::DetectValues`
  - `skip_package_pkg`
    - Similar to `skip_package_ipa` and allows for backwards compatability 
- After `gym` copies Mac app (the `.app` extension file), it will...
  - Get Info.plist path from `Gym.project` (needed for signing)
  - Get entitlements path from `Gym.project` (needed for signing)
  - Call `productbuild` with `: mac_app_installer_cert_name` to generate and sign the `.pkg`
  - Store the `.pkg` path in `SharedValues::PKG_OUTPUT_PATH`

### Catalyst
- `gym` has one new option required for Mac App Store signing
  - `catalyst_platform `
- `Gym::Runner` looks at `: catalyst_platform` to determine which platform to build form
  - This is needed since a Catalyst xcodeproject is **both** iOS and macOS
- `Fastlane::Project` an now detect Catalyst enable projects
- `Gym::DetectValues` will autofill `:destination` with the correct platform based on `: catalyst_platform `

### macOS
```rb
lane :build do
  gym(
    export_method: 'app-store',

    # NEEDED: new option for macOS signing/packaging
    mac_app_installer_cert_name: '3rd Party Mac Developer Installer: Something LLC (ABC1234XYZ)'
  )
  deliver
end
```

### Catalyst macOS
```rb
lane :build do
  gym(
    export_method: 'app-store',

    # NEEDED: new option for selecting platform to build
    catalyst_platform: 'macos',

    # NEEDED: new option for macOS signing/packaging
    mac_app_installer_cert_name: '3rd Party Mac Developer Installer: Something LLC (ABC1234XYZ)'
  )
  deliver
end
```

### Catalyst iOS
```rb
lane :build do
  gym(
    export_method: 'app-store',

    # NEEDED: new option for selecting platform to build
    catalyst_platform: 'ios',
  )
  deliver
end
```

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by janpio** [comment]
Awesome someone finally spent the time to understand what was going on here.

`gym` (the action) actually is called `build_ios_app` now. Do you think there should be some changes (or at least an alias) to that?

**Comment by joshdholtz** [comment]
@janpio lol, the name `build_ios_app` does kind of cause a bit of problem now doesn't it 🤔

**Comment by tkrajacic** [comment]
`build_ios_app` is indicative of the problems macOS developers face. 😀

I mean the upcoming fastlane.ci doesn't even mention macOS plans and their entire premise is that there needs to be a better CI system for mobile development (sure, I agree).
Maddening though is that there are plenty of options for mobile devs available and the real problem is when you start doing macOS development. But I digress…

Thank you @joshdholtz for investing your time in this!

**Comment by fousa** [comment]
@joshdholtz I'm currently testing this PR, I really need this for an app I wrote.

Is there a way I can contribute? I had some issues, but fixed them in the code. I can comment them if that is the best way?

**Comment by fousa** [comment]
Also: great work! 👍

**Comment by joshdholtz** [comment]
@fousa Heyyy, glad you found this PR 😊 If you want to comment them here that would be work or you could branch off my branch and create a PR into my branch but not sure if that is weird or not 😉 But I'm excited to see what you found! Did this end up working for you after you made your fixes?

**Comment by fousa** [comment]
Yep, I got my app onto iTunes Connect, so great!

I'll comment 2 minor issues this evening.

File: gym/lib/gym/runner.rb [hunk 1]
```diff
@@ -192,6 +194,76 @@ def copy_mac_app
...
+      exe_name = File.basename(app_path)
+      pkg_name = "#{exe_name}.pkg"
+      pkg_path = File.join(Gym.config[:output_directory], pkg_name)
+
+      File.delete(pkg_path) if File.exist?(pkg_path)
+
+      # Generates command for signing for distribution
+      codesign_commands = []
+      codesign_commands << "codesign"
+      codesign_commands << "--entitlements #{entitlements_path}"
```
Comment (fousa) [inline gym/lib/gym/runner.rb]: I had an issue where the scheme name contains spaces and this errored.

Fix:
```ruby
codesign_commands << "--entitlements \"#{entitlements_path}\""
```

File: gym/lib/gym/runner.rb [hunk 2]
```diff
@@ -192,6 +194,76 @@ def copy_mac_app
...
+      productbuild_commands << "/Applications"
+      productbuild_commands << "--sign \"#{installation_cert}\""
+      productbuild_commands << "--product \"#{plist_path}\""
+      productbuild_commands << "\"#{pkg_path}\""
+
+      codesign_command = codesign_commands.join(' ')
+      productbuild_command = productbuild_commands.join(' ')
+
+      # Signs application for distribution
+      print_command(command, "Generated Codesign Command") if FastlaneCore::Globals.verbose?
```
Comment (fousa) [inline gym/lib/gym/runner.rb]: Verbose mode crashes.

Fix:
```ruby
print_command(codesign_command, "Generated Codesign Command") if FastlaneCore::Globals.verbose?
```

File: gym/lib/gym/runner.rb [hunk 3]
```diff
@@ -192,6 +194,76 @@ def copy_mac_app
...
+      FastlaneCore::CommandExecutor.execute(command: codesign_command,
+                                          print_all: false,
+                                      print_command: !Gym.config[:silent],
+                                              error: proc do |output|
+                                                ErrorHandler.handle_package_error(output)
+                                              end)
+
+      # Generates package and signs for installation
+      print_command(command, "Generated Productbuild Command") if FastlaneCore::Globals.verbose?
```
Comment (fousa) [inline gym/lib/gym/runner.rb]: Verbose mode crashes.

Fix
```ruby
print_command(productbuild_command, "Generated Productbuild Command") if FastlaneCore::Globals.verbose?
```

**Review by fousa — APPROVED** [review]
Hope this gets into the Fastlane! I would love to have both my iOS as my macOS deployed with Fastlane! 🤘

**Comment by fousa** [comment]
There was one small other thing. When the app file got generated, I had to manually remove the files before I could retry. It seems that the `FileUtils.cp_r` didn't want to overwrite the files.

File: gym/lib/gym/runner.rb (hunk 1 above)
Comment (janpio) [reply gym/lib/gym/runner.rb]: Turned into a suggestion that can just be applied to the code:
```suggestion
      codesign_commands << "--entitlements #{entitlements_path.shellescape}"
```

File: gym/lib/gym/runner.rb (hunk 2 above)
Comment (janpio) [reply gym/lib/gym/runner.rb]: Turned into a suggestion that can just be applied:

```suggestion
      print_command(codesign_command, "Generated Codesign Command") if FastlaneCore::Globals.verbose?
```

File: gym/lib/gym/runner.rb (hunk 3 above)
Comment (janpio) [reply gym/lib/gym/runner.rb]: Turned into a suggestion that can just be applied:
```suggestion
      print_command(productbuild_command, "Generated Productbuild Command") if FastlaneCore::Globals.verbose?
```

**Review by janpio — CHANGES_REQUESTED** [review]
See https://github.com/fastlane/fastlane/pull/12195#issuecomment-393662335

File: gym/lib/gym/options.rb [hunk 4]
```diff
@@ -139,6 +139,16 @@ def self.plain_options
                                      is_string: false,
                                      type: Boolean,
                                      optional: true),
+        FastlaneCore::ConfigItem.new(key: :mac_app_distribution_cert_name,
+                                     env_name: "GYM_MAC_APP_DISTRIBUTION_CERT_NAME",
+                                     description: "Full name of Mac App Distribution Certificate. Example: 3rd Party Mac Developer Application: Your Company (ABC1234XWYZ)",
```
Comment (janpio) [inline gym/lib/gym/options.rb]:
```suggestion
                                     description: "Full name of Mac App Distribution Certificate. Example: `3rd Party Mac Developer Application: Your Company (ABC1234XWYZ)`",
```

File: gym/lib/gym/options.rb [hunk 5]
```diff
@@ -139,6 +139,16 @@ def self.plain_options
...
+                                     is_string: true,
+                                     optional: true),
+        FastlaneCore::ConfigItem.new(key: :mac_app_installer_cert_name,
+                                     env_name: "GYM_MAC_APP_INSTALLER_CERT_NAME",
+                                     description: "Full name of Mac Installer Distribution Certificate. Example: 3rd Party Mac Developer Installer: Your Company (ABC1234XWYZ)",
```
Comment (janpio) [inline gym/lib/gym/options.rb]:
```suggestion
                                     description: "Full name of Mac Installer Distribution Certificate. Example: `3rd Party Mac Developer Installer: Your Company (ABC1234XWYZ)`",
```

**Review by janpio — APPROVED** [review]
Besides what has already been commented and suggested, the code looks great! 
If those are merged and taken care of, this should be good to go in.

File: gym/lib/gym/runner.rb [hunk 6]
```diff
@@ -237,6 +239,76 @@ def copy_mac_app
       app_path
     end
 
+    def build_mac_pkg(app_path)
```
Comment (janpio) [inline gym/lib/gym/runner.rb]: Maybe this should get a better name that reflects "codesign and package"? 
`build` sounds pretty lame her - building happened before, this is finalizing it, right?

**Comment by janpio** [comment]
Open things in this PR:

- Has to be tested again if it works
- https://github.com/fastlane/fastlane/pull/12195#issuecomment-393662335
- https://github.com/fastlane/fastlane/pull/12195#discussion_r329304569
- https://github.com/fastlane/fastlane/pull/12195#issuecomment-377305685
- tests would be nice
- backwards compatibility might need some work
- more real world examples

**Comment by janpio** [comment]
@googlebot I consent

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: fastlane_core/lib/fastlane_core/project.rb [hunk 7]
```diff
@@ -273,6 +273,10 @@ def mac_framework?
       (framework? && build_settings(key: "PLATFORM_NAME") == "macosx")
     end
 
+    def mac_catalyst?
```
Comment (joshdholtz) [inline fastlane_core/lib/fastlane_core/project.rb]: Rename to something like `supports_mac_catalyst?`

File: gym/lib/gym/code_signing_mapping.rb [hunk 8]
```diff
@@ -112,10 +119,13 @@ def detect_configuration_for_archive
       return configuration
     end
 
+    # rubocop:disable Metrics/PerceivedComplexity
     def detect_project_profile_mapping
       provisioning_profile_mapping = {}
       specified_configuration = detect_configuration_for_archive
 
+      is_building_mac_catalyst = Gym.project.mac_catalyst? && Gym.config[:catalyst_platform] == "macos"
```
Comment (joshdholtz) [inline gym/lib/gym/code_signing_mapping.rb]: Add method onto `Gym` module like `Gym. is_building_mac_catalyst_for_mac?` and `Gym. is_building_mac_catalyst_for_ios?` but maybe better

File: gym/lib/gym/code_signing_mapping.rb [hunk 9]
```diff
@@ -88,6 +88,13 @@ def same_platform?(sdkroot)
       when "tvOS"
         destination_sdkroot = ["appletvos"]
       end
+
+      # Catalyst projects will always have an "iphoneos" sdkroot
+      # Need to force a same platform when trying to build as macos
+      if Gym.project.mac_catalyst? && Gym.config[:catalyst_platform] == "macos"
```
Comment (joshdholtz) [inline gym/lib/gym/code_signing_mapping.rb]: Replace with is the new gym module helper

File: gym/lib/gym/detect_values.rb [hunk 10]
```diff
@@ -100,6 +101,28 @@ def self.detect_selected_provisioning_profiles
       end
     end
 
+    # Detects name of a "3rd Party Mac Developer Installer" cert for the configured team id
+    def self.detect_third_party_installer
+      team_id = Gym.config[:export_team_id] || Gym.project.build_settings(key: "DEVELOPMENT_TEAM")
```
Comment (joshdholtz) [inline gym/lib/gym/detect_values.rb]: Need to early return if `: mac_app_installer_cert_name` already set

File: gym/lib/gym/detect_values.rb [hunk 11]
```diff
@@ -111,14 +134,22 @@ def self.min_xcode8?
     # Is it an iOS device or a Mac?
     def self.detect_platform
       return if Gym.config[:destination]
-      platform = if Gym.project.mac?
+      platform = if Gym.project.mac_catalyst?
```
Comment (joshdholtz) [inline gym/lib/gym/detect_values.rb]: Replace with gym module helper

File: gym/lib/gym/generators/package_command_generator_xcode7.rb [hunk 12]
```diff
@@ -160,6 +182,10 @@ def config_content
           hash[:signingStyle] = 'manual'
         end
 
+        if Gym.config[:mac_app_installer_cert_name] && (Gym.project.mac? || Gym.config[:catalyst_platform] == "macos")
```
Comment (joshdholtz) [inline gym/lib/gym/generators/package_command_generator_xcode7.rb]: Replace with gym helper for Mac catalyst guy thing

File: gym/lib/gym/runner.rb [hunk 13]
```diff
@@ -23,7 +23,20 @@ def run
 
       FileUtils.mkdir_p(File.expand_path(Gym.config[:output_directory]))
 
-      if Gym.project.ios? || Gym.project.tvos?
+      # Determine platform to archive
+      if Gym.project.mac_catalyst?
```
Comment (joshdholtz) [inline gym/lib/gym/runner.rb]: Rework this to use gym module helper thing

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 14]
```diff
@@ -1,13 +1,160 @@
...
+        gym_output_path = Gym::Manager.new.work(values)
+        if gym_output_path.nil?
+          UI.important("No output path received from gym")
+          return nil
+        end
+
+        absolute_output_path = File.expand_path(gym_output_path)
+
+        absolute_dsym_path = absolute_output_path.gsub(".ipa", ".app.dSYM.zip") if File.extname(absolute_output_path) == ".ipa"
+        absolute_dsym_path = absolute_output_path.gsub(".pkg", ".dSYM.zip") if File.extname(absolute_output_path) == ".pkg"
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/build_app.rb]: Check for spa and pkg at the end

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 15]
```diff
@@ -1,13 +1,160 @@
...
+            configuration: "Debug",
+            scheme: "MyApp",
+            silent: true,
+            clean: true,
+            output_directory: "path/to/dir", # Destination directory. Defaults to current directory.
+            output_name: "my-app.ipa",       # specify the name of the .ipa file to generate (including file extension)
+            sdk: "iOS 11.1"                  # use SDK as the name or path of the base SDK when building the project.
+          )',
+          'gym              # alias for "build_app"',
+          'build_ios_app    # alias for "build_app"',
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/build_app.rb]: `# alias for "build_app" but only iOS options`
`# alias for "build_app" but only macOS options`

File: deliver/lib/deliver/runner.rb [hunk 16]
```diff
@@ -140,14 +140,14 @@ def prepare_app_icons(options = {})
     # Upload the binary to App Store Connect
     def upload_binary
       UI.message("Uploading binary to App Store Connect")
-      if options[:ipa]
+      if options[:ipa] && ["ios", "appletvos"].include?(options[:platform])
```
Comment (joshdholtz) [inline deliver/lib/deliver/runner.rb]: THIS BREAKS... only check for platform if both IPA and PKG exist

**Review by joshdholtz — COMMENTED** [review]
I WANNA REQUEST MY OWN CHANGES

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 17]
```diff
@@ -1,13 +1,160 @@
...
+      IPA_OUTPUT_PATH ||= :IPA_OUTPUT_PATH
+      PKG_OUTPUT_PATH ||= :PKG_OUTPUT_PATH
+      DSYM_OUTPUT_PATH ||= :DSYM_OUTPUT_PATH
+      XCODEBUILD_ARCHIVE ||= :XCODEBUILD_ARCHIVE # originally defined in XcodebuildAction
+    end
+
+    class BuildAppAction < Action
+      # rubocop:disable Metrics/PerceivedComplexity
+      def self.run(values)
+        require 'gym'
```
Comment (snatchev) [inline fastlane/lib/fastlane/actions/build_app.rb]: why are these required here and not at the top, on-load?

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 18]
```diff
@@ -1,13 +1,160 @@
...
-      #####################################################
+    module SharedValues
+      IPA_OUTPUT_PATH ||= :IPA_OUTPUT_PATH
+      PKG_OUTPUT_PATH ||= :PKG_OUTPUT_PATH
+      DSYM_OUTPUT_PATH ||= :DSYM_OUTPUT_PATH
+      XCODEBUILD_ARCHIVE ||= :XCODEBUILD_ARCHIVE # originally defined in XcodebuildAction
+    end
+
+    class BuildAppAction < Action
+      # rubocop:disable Metrics/PerceivedComplexity
```
Comment (snatchev) [inline fastlane/lib/fastlane/actions/build_app.rb]: I think we can refactor this method a little bit to avoid this disable. Let's make 2 methods for the big conditional branch that accept a `export_options` parameter, and return the transformed `export_options`.

We can probably also rework this to exit early with `show_xcode_9_warning` in the correct case.

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 19]
```diff
@@ -1,13 +1,160 @@
...
+        gym_output_path = Gym::Manager.new.work(values)
+        if gym_output_path.nil?
+          UI.important("No output path received from gym")
+          return nil
+        end
+
+        absolute_output_path = File.expand_path(gym_output_path)
+
+        absolute_dsym_path = absolute_output_path.gsub(/.ipa$/, ".app.dSYM.zip") if File.extname(absolute_output_path) == ".ipa"
+        absolute_dsym_path = absolute_output_path.gsub(/.pkg$/, ".dSYM.zip") if File.extname(absolute_output_path) == ".pkg"
```
Comment (snatchev) [inline fastlane/lib/fastlane/actions/build_app.rb]: This file type check e.g. `File.extname(absolute_output_path) == ".ipa"` is done twice. Move these under
the check below.

File: fastlane/lib/fastlane/actions/build_app.rb [hunk 20]
```diff
@@ -1,13 +1,160 @@
...
+
+        # This might be the mac app path, so we don't want to set it here
+        # https://github.com/fastlane/fastlane/issues/5757
+        if absolute_output_path.include?(".ipa")
```
Comment (snatchev) [inline fastlane/lib/fastlane/actions/build_app.rb]: prefer `File.extname` instead of substring check.

File: gym/lib/gym/generators/package_command_generator_xcode7.rb [hunk 21]
```diff
@@ -79,6 +79,28 @@ def ipa_path
         Gym.cache[:ipa_path]
       end
 
+      def pkg_path
+        unless Gym.cache[:pkg_path]
```
Comment (snatchev) [inline gym/lib/gym/generators/package_command_generator_xcode7.rb]: remove conditional
```
pkg_path = Gym.cache[:pkg_path]
return pkg_path if pkg_path
```
use local variable instead of cache-lookup.

**Review by snatchev — CHANGES_REQUESTED** [review]
just a few suggestions :)

File: deliver/lib/deliver/runner.rb [hunk 22]
```diff
@@ -140,14 +140,23 @@ def prepare_app_icons(options = {})
...
+
+      if upload_ipa
         package_path = FastlaneCore::IpaUploadPackageBuilder.new.generate(
           app_id: options[:app].apple_id,
           ipa_path: options[:ipa],
           package_path: "/tmp",
           platform: options[:platform]
         )
-      elsif options[:pkg]
+      elsif upload_pkg
```
[the remark this answers is not in the document]
Comment (joshdholtz) [reply deliver/lib/deliver/runner.rb]: That's prevented in the options verify block 😊

File: fastlane/lib/fastlane/actions/build_app.rb (hunk 17 above)
Comment (joshdholtz) [reply fastlane/lib/fastlane/actions/build_app.rb]: All the actions get loaded when fastlane starts and we really only want the action to require things it needs when it runs (otherwise the initial load of fastlane would take longer and load all the things for all the actions)

File: fastlane/lib/fastlane/actions/build_app.rb (hunk 18 above)
Comment (joshdholtz) [reply fastlane/lib/fastlane/actions/build_app.rb]: This is no longer complex now that we reworked the ipa/pkg/dsym logic below 😊

**Review by snatchev — APPROVED** [review]
LGTM :)
